File: flask_app/app/controllers/user_controller.py
from flask import request
from flask_jwt_extended import jwt_required
from app.services.user_service import UserService
from app.utils.response import success_response, error_response
from app.utils.decorators import admin_required


class UserController:
    @staticmethod
    @admin_required()
    def get_all():
        page = request.args.get("page", default=1, type=int)
        per_page = request.args.get("per_page", default=10, type=int)

        filters = {"status": request.args.get("status")}

        filters = {k: v for k, v in filters.items() if v is not None}

        users, total = UserService.get_all(page, per_page, filters)
        return success_response(
            "Users retrieved",
            {
                "items": [u.to_dict() for u in users],
                "total": total,
                "page": page,
                "per_page": per_page,
            },
        )

    @staticmethod
    @jwt_required()
    def get_one(user_id):
        user = UserService.get_by_id(user_id)
        if not user:
            return error_response("User not found", 404)
        return success_response("User found", user.to_dict())

    @staticmethod
    @jwt_required()
    def update(user_id):
        data = request.get_json()
        updated_user = UserService.update(user_id, data)
        if not updated_user:
            return error_response("User not found", 404)
        return success_response("User updated", updated_user.to_dict())

    @staticmethod
    @admin_required()
    def delete(user_id):
        deleted = UserService.soft_delete(user_id)
        if not deleted:
            return error_response("User not found", 404)
        return success_response("User deleted successfully")

File: java_membership_api/src/main/java/com/example/membership_api/controller/UserController.java
package com.example.membership_api.controller;

import java.util.Map;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.membership_api.annotations.AdminOnly;
import com.example.membership_api.dto.UserPage;
import com.example.membership_api.model.User;
import com.example.membership_api.service.UserService;

@RestController
@RequestMapping("/users")
public class UserController {

    private static final Logger logger = LoggerFactory.getLogger(UserController.class);

    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping("")
    @AdminOnly
    public ResponseEntity<UserPage> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int perPage,
            @RequestParam Map<String, String> filters) {

        logger.debug("Controller: getAll called with page {}, perPage {}, filters {}", page, perPage, filters);

        // Remove pagination params from filters map so DAO only gets actual filters
        filters.remove("page");
        filters.remove("perPage");

        UserPage userPage = userService.getAll(page, perPage, filters);
        return ResponseEntity.ok(userPage);
    }

    @GetMapping("/{userId}")
    public ResponseEntity<User> getOne(@PathVariable("userId") String userId) {
        logger.debug("Controller: getOne called with userId {}", userId);

        Optional<User> userOpt = userService.getById(userId);
        return userOpt
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PutMapping("/{userId}")
    public ResponseEntity<User> update(@PathVariable("userId") String userId,
                                       @RequestBody Map<String, Object> updates) {
        logger.debug("Controller: update called with userId {}, updates {}", userId, updates);

        Optional<User> updatedUser = userService.update(userId, updates);
        return updatedUser
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    // DELETE /api/users/{user_id}
    @AdminOnly
    @DeleteMapping("/{userId}")
    public ResponseEntity<Void> delete(@PathVariable("userId") String userId) {
        logger.debug("Controller: delete called with userId {}", userId);

        boolean deleted = userService.softDelete(userId);
        if (deleted) {
            return ResponseEntity.noContent().build();
        } else {
            return ResponseEntity.notFound().build();
        }
    }
}
